### DISCUSSION TITLE
Understanding performance discrepancies between IPOPT logs and actual elapsed time in opti-stack

### DISCUSSION THREAD

**Discussion by chclerget** [casadi/casadi#3246] [Q&A]
Hi,

I'm using Opti to solve various kind of optimization problems with IPOPT. I've noticed that in some cases, the actual time spent to formulate the problem and solve it is much longer than what seems to be implied by IPOPT's logs.

For instance, consider this large but simple regression problem where I identify a linear model with two parameters

---------------

import casadi
import numpy as np

#### Generating data from linear process
n = 10 ** 6

X = np.random.normal(loc=0.0, scale=1.0, size=(n,2))

theta = np.array([[1], [0.5]])

y = X @ theta + np.random.normal(loc=0.0, scale=0.1, size=(n, 1))

#### Estimating theta with casadi
opti = casadi.Opti()

theta_est = opti.variable(2, 1)

y_est = X @ theta_est

loss = casadi.sum1((y_est - y) ** 2)

opti.minimize(loss)

p_opts = {'print_time': 0}
s_opts = {}
        
opti.solver('ipopt', p_opts, s_opts)

start_solve = time.process_time()

sol = opti.solve()

---------------

IPOPT, returns the following logs

Number of objective function evaluations             = 2
Number of objective gradient evaluations             = 2
Number of equality constraint evaluations            = 0
Number of inequality constraint evaluations          = 0
Number of equality constraint Jacobian evaluations   = 0
Number of inequality constraint Jacobian evaluations = 0
Number of Lagrangian Hessian evaluations             = 1
Total CPU secs in IPOPT (w/o function evaluations)   =      0.036
Total CPU secs in NLP function evaluations           =      0.136

This seems to imply that the CPU time required to solve the problem is ~0.17s. However, when profiling the second block of code starting at the declaration of the opti variable with time.process_time(), I find that the total time spent formulating and solving the problem with Opti is actually ~1.4s (out of that ~1s is spent in opti.solve()).

My questions are: is such a large discrepency expected, where is the overhead coming from and am I doing something wrong in the way I use casadi?

**Comment by jgillis** [comment]
The first call to opti.solve() does both construction and numerical evaluation. Only the latter is timed. Construction entails symbolic computations (source code transforming algorithmic differentiation on the expression graphs) and memory allocations. Subsequent calls (e.g. when you have modified initial guesses or parameters) only perform numerical evaliation.

**Reply by chclerget** [reply]
Hi, thanks you very much for clarifying this.

Is there anything I could do to lower that overhead of the problem construction or do you think this is just "what it is"? The example seems basic enough that there shouldn't be much nuance but I'm not really familiar with casadi's internal machinery so I might be doing something stupid.

Not a completely fair comparison, but sklearn solves that linear regresion in ~0.05s.

**Reply by jgillis** [reply]
Is the question in the context of publishing a comparison of software tools? There are some tricks possible, but in general we never optimized code for fast construction time, only evaluation time. In 'production', you'd just have the construction time at startup, followed by an infinite loop of numerical evaluations. CasADi was made with parametric nonlinear constrained optimization in mind, and some trade-offs may well adversily affect simple regressiom problems. For instance, we don't even have a dense matrix type, everything is sparse. Also the overhead in our numpy interface may ne substantial here.

**Reply by chclerget** [reply]
That's good context, thanks.

My applications are not in optimal control/MPC, casadi is rather my go-to AML for optimization in a Python environment. Most of my use cases are either ML-like (implementing unconventional regressions) or offline signal processing/parameter estimation. The overwhelming majority of my opti instances only get solved once.

Happy to DM if you want to chat more. I've been pretty happy with casadi so far, but might not be the best tool for me.

**Reply by chclerget** [reply]
Thanks @jgillis. Coming back to this, is there another Python package that you think would be better suited for the formulation of generic optimization problems with more offline applications in mind?
